src/setup-beam.js

```diff
@@ -592,6 +596,327 @@ function debugLog(groupName, message) {
+            } else {
+              url = `https://github.com/erlang/rebar3/releases/download/${versionSpec}/rebar3`
+            }
+
+            return url
+          },
+          extract: async () => ['file', 'rebar3'],
+          postExtract: async (cachePath) => {
+            const bindir = path.join(cachePath, 'bin')
+            const oldPath = path.join(cachePath, 'rebar3')
+            const newPath = path.join(bindir, 'rebar3')
+            fs.mkdirSync(bindir)
+            fs.renameSync(oldPath, newPath)
+            fs.chmodSync(newPath, 0o755)
+          },
+          reportVersion: () => {
+            const cmd = 'rebar3'
+            const args = ['version']
+
+            return [cmd, args]
+          },
+        },
+        win32: {
+          downloadToolURL: () => {
+            let url
+            if (versionSpec === 'nightly') {
+              url = 'https://s3.amazonaws.com/rebar3-nightly/rebar3'
+            } else {
+              url = `https://github.com/erlang/rebar3/releases/download/${versionSpec}/rebar3`
+            }
+
+            return url
+          },
+          extract: async () => ['file', 'rebar3'],
+          postExtract: async (cachePath) => {
+            const bindir = path.join(cachePath, 'bin')
+            const oldPath = path.join(cachePath, 'rebar3')
+            fs.mkdirSync(bindir)
+            fs.chmodSync(oldPath, 0o755)
+
+            const ps1Filename = path.join(bindir, 'rebar3.ps1')
+            fs.writeFileSync(ps1Filename, `& escript.exe ${oldPath} \${args}`)
+
+            const cmdFilename = path.join(bindir, 'rebar3.cmd')
+            fs.writeFileSync(
+              cmdFilename,
+              `@echo off\r\nescript.exe ${oldPath} %*`,
+            )
+          },
+          reportVersion: () => {
+            const cmd = 'rebar3.cmd'
+            const args = ['version']
+
+            return [cmd, args]
+          },
+        },
+      }
+      break
+    default:
+      throw new Error(`no installer for ${toolName}`)
+  }
+
+  await installTool({ toolName, versionSpec, installOpts })
+}
+
+async function installTool(opts) {
+  const { toolName, versionSpec, installOpts } = opts
+  const platformOpts = installOpts[process.platform] || installOpts.all
+  let cachePath = tc.find(toolName, versionSpec)
+
+  core.debug(`Checking if ${installOpts.tool} is already cached...`)
+  if (cachePath === '') {
+    core.debug("  ... it isn't!")
+    const downloadToolURL = platformOpts.downloadToolURL()
+    const file = await tc.downloadTool(downloadToolURL)
+    const [targetElemType, targetElem] = await platformOpts.extract(file)
+
+    if (targetElemType === 'dir') {
+      cachePath = await tc.cacheDir(targetElem, toolName, versionSpec)
+    } else if (targetElemType === 'file') {
+      cachePath = await tc.cacheFile(file, targetElem, toolName, versionSpec)
+    }
+  } else {
+    core.debug(`  ... it is, at ${cachePath}`)
+  }
```

Comment:
Along with the comment block shown above, this is the function that implements the generic "behaviour" used, at this moment, for 4 tools in 2 OS types. This has been tested with the addition of Gleam and Rebar3 to `tool-cache` and retrofitted for Erlang/OTP and Elixir.